storage/innobase/handler/i_s.cc

```diff
@@ -6873,6 +7106,96 @@ static int i_s_innodb_tablespaces_fill_table(THD *thd, TABLE_LIST *tables,
   }
 
   heap = mem_heap_create(1000);
+
+  if (dd_access_through_server()) {
+    return for_all_innodb_objects_in_dd<dd::Tablespace>(
+        thd, [&heap, &tables](THD *thd, const dd::Tablespace &tablespace) {
+          const auto &se_private_data = tablespace.se_private_data();
+
+          space_id_t space;
+          if (se_private_data.get(dd_space_key_strings[DD_SPACE_ID], &space)) {
+            assert(false);
+            mem_heap_free(heap);
+            return 1;
+          }
+
+          if (space == 0) return 0;
+
+          auto *const name = static_cast<char *>(
+              mem_heap_zalloc(heap, tablespace.name().length() + 1));
+          memcpy(name, tablespace.name().c_str(),
+                 tablespace.name().length() + 1);
+
+          const auto &properties = tablespace.options();
+
+          uint32_t flags;
+          if (se_private_data.get(dd_space_key_strings[DD_SPACE_FLAGS],
+                                  &flags)) {
+            assert(false);
+            mem_heap_free(heap);
+            return 1;
```

Comment:
return 0 to minic old innodb behavior? https://github.com/facebook/mysql-5.6/blob/fb-mysql-8.0.28/storage/innobase/dict/dict0dd.cc#L5453

Reply:
The old InnoDB behavior is different in different callstack levels: the lower `dd_process_dd_tablespaces_rec` function that you linked to indeed returns false on error, true on success, but here we are in `i_s_innodb_tablespaces_fill_table`, which returns 0 on success, 1 on error.

Actually I see that to fully mimic the old behavior the error at this point has to be silently eaten